cli/cppcheckexecutor.cpp

```diff
@@ -71,6 +72,117 @@
 #endif
 
 namespace {
+    class SarifReport {
+    public:
+        void addFinding(ErrorMessage msg) {
+            mFindings.push_back(std::move(msg));
+        }
+
+        picojson::array serializeRules() const {
+            picojson::array ret;
+            std::set<std::string> ruleIds;
+            for (const auto& finding : mFindings) {
+                if (ruleIds.insert(finding.id).second) {
+                    picojson::object rule;
+                    rule["id"] = picojson::value(finding.id);
+                    picojson::object shortDescription;
+                    shortDescription["text"] = picojson::value(finding.shortMessage());
+                    rule["shortDescription"] = picojson::value(shortDescription);
+                    ret.emplace_back(rule);
+                }
+            }
+            return ret;
+        }
+
+        static picojson::array serializeLocations(const ErrorMessage& finding) {
+            picojson::array ret;
+            for (const auto& location : finding.callStack) {
+                picojson::object physicalLocation;
+                picojson::object artifactLocation;
+                artifactLocation["uri"] = picojson::value(location.getOrigFile(false));
```

Comment:
Not sure the output of `location.getOrigFile(false)`, but it should be a file name relative to the root of the project/repository for best results. If it is an absolute file path, you may want to set the `workingDirectory.uri` property. See [Specifying the root for source files](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#specifying-the-root-for-source-files).

Reply:
Cppcheck does not know anything about project path. I will expect that users ensure that the filepaths are relative to the project path. But yes the `getfile` is better than the `getOrigFile` then, thanks!